### PR TITLE
plug leaks on finalize

### PR DESCRIPTION
**Pull request by ggouaillardet**

### CODE REVIEW

**Comment by ggouaillardet** [comment]
Refs #2166 

@jsquyres this fixes several memory leaks
if i configure with ```--disable-dlopen```, i can run
```
valgrind --leak-check=full --show-leak-kinds=all ./hello_c
```
and it shows no Open MPI memory leaks 
/* i see two memory leaks, one in ```libibverbs``` and one in ```libinfinipath``` */

**Comment by lanl-ompi** [comment]
Test FAILed.

**Comment by lanl-ompi** [comment]
Test FAILed.

**Comment by ibm-ompi** [comment]
Build Failed with GNU compiler! Please review the log, and get in touch if you have questions.

 Gist: https://gist.github.com/2dc4f42e6f3820b9f92169e224ffaa7f

**Comment by ibm-ompi** [comment]
Build Failed with PGI compiler! Please review the log, and get in touch if you have questions.

 Gist: https://gist.github.com/e14814a6b03f72d5f33b1be4588fb2c0

**Comment by bosilca** [comment]
This is AWESOME!!! It might even be the first time we are memory leak free. Kudos @ggouaillardet !

**Comment by lanl-ompi** [comment]
Test FAILed.

**Comment by lanl-ompi** [comment]
Test FAILed.

**Comment by ibm-ompi** [comment]
Build Failed with GNU compiler! Please review the log, and get in touch if you have questions.

 Gist: https://gist.github.com/83523256687e587cca78231d7e057b16

**Comment by ibm-ompi** [comment]
Build Failed with PGI compiler! Please review the log, and get in touch if you have questions.

 Gist: https://gist.github.com/1103b58d19aa59d9054d34f54c7e66c0

**Comment by rhc54** [comment]
@ggouaillardet Have the PMIx things been backported yet? Are you sure we want a "none" sec module in PMIx? Sounds like a security leak, yes?

**Comment by ggouaillardet** [comment]
@rhc54 i made pmix/master#170 and it has not been merged yet.
then i will make the PR for the PMIx release branch(es)

the ```sec/none``` thing is not relevant here and should have not been included.
this is a band-aid to "fix" #2147 
```fakeroot``` preload a library to trick the process into believing they are root, but fails to trick
```getsockopt (peer->sd, SOL_SOCKET, SO_PEERCRED, &ucred, &crlen);```
hence the failure.
i did not implement this as a normal module (e.g. if all modules fails, fall back to ```sec/none```).
```sec/none``` is used if and only if it is the only module requested by the ```PMIX_SECURITY_MODE``` environment variable.
obviously, it gives the option to disable the security, but the security can only be disabled intentionally.
i am still waiting for the user feedback on the need for ```fakeroot```
if it is not critical to him, i plan to simply drop this commit.

**Comment by rhc54** [comment]
I have no issue with the sec/none component under those conditions. It won't allow you to connect to the system PMIx server, but would allow connections within an mpirun-executed job if the user so directed, and that's fine.

+1 on George's comment!!

**Comment by rhc54** [comment]
BTW: why is this marked as DNM? Can we go ahead and merge it?

**Comment by ggouaillardet** [comment]
i was plugging these leaks on finalize when i saw @jsquyres opened #2166 and i pushed by branch to avoid the job being done twice. there has been a few iterations since (```--enable-dlopen``` support, ```OS X``` support) and very little testing on real apps, hence the "do not merge" label.

i made #2193 to back port the ```pmix3x``` fixes from upstream, and once it is merged, i will push the "safest" commits from this PR and see how things go (e.g. check MTT results)
my plan is then to push the other commits one or two steps at a time in order to let mtt a chance to catch anything that might go wrong.

**Comment by rhc54** [comment]
I'm not wild about us porting patches from libevent into OMPI. This is something we have never done before as we want to stick with a stable release. Please don't commit that part until we have a chance to discuss it - I'm not sure we want libevent_global_shutdown being invoked by a library.

**Comment by ggouaillardet** [comment]
ok, i will not merge this for now.

fwiw, 2.0.22 is the latest stable release of ```libevent```, and that was two years ago.
my main motivation is to be able to automatically detect memory leaks. unless we build an exception for "known" leaks such as global stuff from ```libevent```, ```libevent_global_shutdown()``` has to be invoked to plug some leaks.
i do not see this as an issue for the __embedded__ ```libevent```.

**Comment by rhc54** [comment]
I grok the embedded restriction - I'm more concerned about maintainability. We try to avoid customizing these imported libraries, though I admit we have cherry-picked ROMIO over the years. I'd just like to ensure others are on-board with starting to do that with libevent so we can define the extent to which we do so. For example, IBM opened a ticket on libevent recently to express desires for a new stable release as there are bugs which have been fixed up there - do we want to cherry-pick those too?

@gpaulsen @jsquyres @bosilca Opinions?

**Comment by jsquyres** [comment]
Does the newest libevent have these fixes?

**Comment by rhc54** [comment]
libevent master does, as does 2.1beta - but no stable release does, and we have no firm date for when that might occur

**Comment by ibm-ompi** [comment]
Build Failed with GNU compiler! Please review the log, and get in touch if you have questions.

 Gist: https://gist.github.com/49f2e18c8939a94d4fc80246272e7234

**Comment by ggouaillardet** [comment]
:bot:mellanox:retest

**Comment by gpaulsen** [comment]
Libevent has now released a new stable version: 2.1.8.  IBM is interested in building an external component to interface with libevent 2.1.8, and test that.  We need to test it well before committing to upgrading.

**Comment by jjhursey** [comment]
@gpaulsen Does libevent `2.1.8` break the current libevent external component? Was there an API change? Or do we just need to check to see if the current external component works with that release?

**Comment by gpaulsen** [comment]
The libevent external component is very version specific.  I'm working to create a new external component for the libevent 2.1.8, but am swamped these next two weeks.

**Comment by rhc54** [comment]
Before spending a ton of time on libevent external components, it might be wise to circle around with @hjelmn about the investigation to replace OMPI's use of libevent

**Comment by gpaulsen** [comment]
Oh cool.  I'll be sure to ping him about that.

**Comment by ibm-ompi** [comment]
The IBM CI (PGI Compiler) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/c201bdea04acf807ac1a2f532277b977

**Comment by bosilca** [comment]
Any hope to rebase this and get it into master ?

**Comment by ggouaillardet** [comment]
@bosilca most changes were brought into master over time.

i rebased the PR (and plugged a few more leaks).
when pmix/master#364 is merged and brought back into master
```
valgrind --leak-check=full  ./hello_c
```
will show no leaks, though a more thorough
```
valgrind --leak-check=full  --show-leak-kinds=all ./hello_c
```
still show unplugged leaks.

note some leaks can be plugged by invoking ```libevent_global_shutdown()```, and this only occurs when the
```event_libevent2022_global_shutdown``` MCA param is ```true```

i still need to cleanup the last commit.
for now, and imho, there is little benefit from merging this PR

**Comment by hppritcha** [comment]
closing due to lack of activity.  reopen if this will become active again.

**Comment by bosilca** [comment]
We need to get at least some of these changes in master. I'll work with @ggouaillardet on it.

**Comment by ggouaillardet** [comment]
@bosilca I rebased and updated this PR

**Comment by bosilca** [comment]
@ggouaillardet I would drop the changes to libevent. My understanding of the discussions at the F2F meeting was that we don't want to maintain our own version of libevent, and instead we would like to rely on the one available on the system. Thus, these changes are better off pushed upstream in the libevent repo.

**Comment by ggouaillardet** [comment]
@bosilca the `libevent` change was cherry-picked from its upstream repo (e.g. backported to 2.0.22).
So a better option would be to update the embedded `libevent` to a newer version.

File: ompi/mca/pml/ob1/pml_ob1_isend.c [hunk 1]
```diff
@@ -268,6 +269,16 @@ int mca_pml_ob1_send(const void *buf,
         rc = sendreq->req_send.req_base.req_ompi.req_status.MPI_ERROR;
     }
 
+    if (sendreq->req_send.req_base.req_pml_complete) {
+        /* make buffer defined when the request is compeleted,
```
Comment (bosilca) [inline ompi/mca/pml/ob1/pml_ob1_isend.c]: typo on completed.

File: ompi/mca/pml/ob1/pml_ob1_isend.c [hunk 2]
```diff
@@ -268,6 +269,16 @@ int mca_pml_ob1_send(const void *buf,
         rc = sendreq->req_send.req_base.req_ompi.req_status.MPI_ERROR;
     }
 
+    if (sendreq->req_send.req_base.req_pml_complete) {
```
Comment (bosilca) [inline ompi/mca/pml/ob1/pml_ob1_isend.c]: Why here and not in mca_pml_ob1_send_request_fini, a function called by all request completions ?

File: ompi/mca/topo/treematch/topo_treematch_dist_graph_create.c [hunk 3]
```diff
@@ -683,7 +683,7 @@ int mca_topo_treematch_dist_graph_create(mca_topo_base_module_t* topo_module,
                 free(aff_mat->sum_row);
                 free(aff_mat);
                 tm_free_solution(sol);
-                tm_free_tree(comm_tree);
+                free(comm_tree);
```
Comment (bosilca) [inline ompi/mca/topo/treematch/topo_treematch_dist_graph_create.c]: This structure has been allocated with treematch functions, I think it is safer to release it with treematch functions. From a programming perspective the arguments should have been ** as the structure is really freed, but the structure is correctly released (except for "dumb" trees in free_non_constraint_tree).

@GuillaumeMercier can you clarify this ?

File: ompi/mca/topo/treematch/treematch/tm_tree.h [hunk 4]
```diff
@@ -8,7 +8,6 @@ void display_tab(double **tab,int N);
 void set_node(tm_tree_t *node,tm_tree_t ** child, int arity,tm_tree_t *parent,
 	      int id,double val,tm_tree_t *tab_child,int depth);
 
-
```
Comment (bosilca) [inline ompi/mca/topo/treematch/treematch/tm_tree.h]: This file is a copy from treematch. I would preserve it as much as possible.

**Comment by rhc54** [comment]
@ggouaillardet is any of this still relevant??

**Comment by ggouaillardet** [comment]
@rhc54 yes it is.

I duplicated half the commits into #5624 and will commit once reviewed (and will remove the committed commits from this PR) as soon as possible.

**Comment by AboorvaDevarajan** [comment]
Can one of the admins verify this patch?

**Comment by gpaulsen** [comment]
@ggouaillardet This PR is about a year old now.  Is there any value to keeping it open at this point?  Can it be rebased, or is it easier to start from scratch, now a year later?

**Comment by gpaulsen** [comment]
@awlauria, There is a lot of good memory leak plugs here in this PR.  We should try to resurrect this.

**Comment by ggouaillardet** [comment]
@gpaulsen I did rebase the PR, it compiles but I did not gave it a run.

**Comment by awlauria** [comment]
@ggouaillardet it looks like you forgot the `-s` on your last checkpoint commit. Did you want to fix - or should I.

 I can pick up the pr and give it some runs.

**Comment by awlauria** [comment]
Re PR'ing onto master (#7501 ) after addressing the comments from @bosilca . There may have been things I missed.

**Comment by gpaulsen** [comment]
Closing this PR as Austen's been taking this forward in #7501.

**Pull request closed without merging**